chainstate-types/src/locator.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): L. Kuklinek
+
+use common::chain::block::Block;
+use common::primitives::Id;
+use serialization::{Decode, Encode};
+
+/// Locator is a list of block IDs at exponentially increasing distance from the tip
+#[derive(PartialEq, Eq, Clone, Debug, Encode, Decode)]
+pub struct Locator(Vec<Id<Block>>);
+
+impl Locator {
+    /// A new empty locator
+    pub fn new_empty() -> Locator {
+        Locator(Vec::new())
+    }
+
+    /// A new locator. Only chainstate implementation can create one
+    pub fn new(entries: Vec<Id<Block>>) -> Locator {
+        Locator(entries)
+    }
+
+    /// Get iterator over locator entries
+    pub fn iter(&self) -> impl Iterator<Item = &Id<Block>> + ExactSizeIterator {
+        self.0.iter()
+    }
+
+    /// Get number of entries in the locator
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// COnvert into a vector.
+    pub fn into_vec(self) -> Vec<Id<Block>> {
+        self.0
+    }
+}
+
+impl std::ops::Index<usize> for Locator {
```

Comment:
To be honest I'd rather remove this as it is currently only used in tests. I just did not bother figuring out what a better approach would be for the time of being.

I had `FromIterator` but it did not play nice with the way the locator is actually created in the code, where the iterator goes over `Result<Option<Id<Block>>>` rather than `Id<Block>`. Again it may be possible to solve but I am now sufficiently busy with other stuff.